HIP-3.md

```diff
@@ -0,0 +1,223 @@
+- There are no refunds for overpayments. Because all data and calculation happens off chain, when a payment is made on the origin chain there is no way to determine if a message has been overpaid for. After a message has been relayed, the relayer could potentially refund overpayments, but this incurs a transaction fee. It's also not very clear which address should receive the refund - this would need to be specified explicitly when making the gas payment.
+- It's hard for applications to pay for message fees on their user's behalf. Some contracts are willing to keep a native token balance that they'll use to pay for message fees. However if the fee amount is calculated off-chain and passed into the contract by a user, there's nothing stopping a user from specifying a very high fee and depleting the contract's balance. This approach of contracts paying for fees on a user's behalf is one way of supporting callback messages.
+- Developers are required to use the Hyperlane SDK to estimate correct payments.
+- A transaction won't revert if there is insufficient payment for a message that is being dispatched. With on chain fee quoting, the transaction can be reverted if there is insufficient payment, creating a very strong feedback loop to users. Sending a message should “just work.”
+- Generally, it's a rough user experience. Application developers have an easier time reasoning about and dealing with on-chain data, and want to have certainty that messages will be processed. Requiring off-chain fee calculation limits the composability of Hyperlane.
+
+### **Tech Spec**
+
+Paying for fees can be done by paying an `InterchainGasPaymaster` contract relating to an off-chain relayer that will enter a social contract to process a message. This contract is specific to a relayer operator who has the ability to claim fees paid to the contract.
+
+At a high level, the `InterchainGasPaymaster` contract should accept the message ID of a Hyperlane message that has been dispatched, information about how much gas the relayer should use when processing the message, and payment in the form of native tokens.
+
+Below is the proposed generic interface for an `InterchainGasPaymaster`. Note that implementations of an IGP may place differring semantic meaning to each of the parameters.
+
+```solidity
+interface IInterchainGasPaymaster {
+    /**
+     * @notice Accepts native tokens to pay for the destination gas cost of a
+     *   message. Implementations may or may not require specific payment amounts.
+     * @param _messageId The message ID.
+     * @param _destinationDomain The destination domain of the message.
+     * @param _gas An amount of gas.
+     * @param _refundAddress The address to refund any overpayment to.
+     */
+    function payForGas(
```

Comment:
Should there be another implementation of this that omits `_gas` and `_refundAddress`? For destination chains where on-chain fee quoting isn't happening

Reply:
> this would be in a scenario where someone is using our IGP that does perform on-chain fee quoting, but they are calling payForGas with a domain that doesn't have any oraclized fees / gas prices?

Correct. Basically I don't want to add additional requirements for supporting a new chain. Relaying should be able to work in the absence of on-chain oraclization

Reply:
To underscore, making deploying Hyperlane to new chains as easy as possible is one of the most important things we can do, and it's important that the gas payment protocol aligns with that goal